### PR TITLE
Add new parameter named auto_complete_list

### PR DESCRIPTION
**Pull request by dcherniev**
Updated HMI_API.xml and MOBILE_API.xml with auto_complete_list field;
Added auto_complete_list field in smart_object_keys;
Added checks in set_global_properties_request to check the transfering of new field;
Added keyboardPropertiesDefault structure in HMI_capabilities.json;
Added unit tests to cover changes from CRQ.

### CODE REVIEW

**Comment by dcherniev** [comment]
Please review: @LuxoftAKutsan, @TMelnyk, @okozlovlux, @LevchenkoS, @dtrunov

File: src/components/application_manager/test/commands/mobile/set_global_properties_test.cc [hunk 1]
```diff
@@ -63,64 +66,79 @@ using ::testing::Return;
...
+      , mock_app_(CreateMockApp()) {
+    Mock::VerifyAndClearExpectations(&mock_message_helper_);
+    ON_CALL(app_mngr_, application(_)).WillByDefault(Return(mock_app_));
     ON_CALL(*mock_app_, app_id()).WillByDefault(Return(kConnectionKey));
     ON_CALL(app_mngr_, hmi_interfaces())
         .WillByDefault(ReturnRef(hmi_interfaces_));
   }
 
-  void TearDown() OVERRIDE {
+  ~SetGlobalPropertiesRequestTest() {
```
Comment (okozlovlux) [inline src/components/application_manager/test/commands/mobile/set_global_properties_test.cc]: @dcherniev , why did you change `teardown` to dtor?

**Comment by TMelnyk** [comment]
@dcherniev , please do NOT change date or version at MOBILE_API

File: src/components/application_manager/test/commands/mobile/set_global_properties_test.cc (hunk 1 above)
Comment (dcherniev) [inline src/components/application_manager/test/commands/mobile/set_global_properties_test.cc]: @okozlovlux Judging by the content of SetUp and TearDown functions I decided that the same logic can be done by constructor and destructor, so I moved content of those functions to ctor and dtor.

File: src/components/application_manager/test/commands/mobile/set_global_properties_test.cc [hunk 2]
```diff
@@ -63,64 +66,79 @@ using ::testing::Return;
...
+    (*msg_vr)[strings::msg_params] = msg_params;
+
+    return msg_vr;
+  }
+
   void ResultCommandExpectations(MessageSharedPtr msg,
                                  const std::string& info) {
-    EXPECT_EQ((*msg)[am::strings::msg_params][am::strings::success].asBool(),
-              true);
+    EXPECT_EQ((*msg)[strings::msg_params][strings::success].asBool(), true);
```
Comment (dcherniev) [inline src/components/application_manager/test/commands/mobile/set_global_properties_test.cc]: @okozlovlux Fixed, please check e3d3f1b.

**Comment by dcherniev** [comment]
@TMelnyk Removed date change, made in MOBILE_API, please check 44c2853. But what about HMI_API, is it ok to change date in there?

**Comment by TMelnyk** [comment]
We have an article at Confluence about API versioning
PLease check once more
HMI_API is ok
MOBILE_API - do not change date and version if CRQ has NO info about it

**Comment by dcherniev** [comment]
@TMelnyk Ok, thank you for the hint, I have also updated the minor version of "Common" interface, please check c675166.

**Comment by TMelnyk** [comment]
Common at HMI_API ?
It`s ok, please read article at Confluence

**Comment by dcherniev** [comment]
@TMelnyk Yep it is in HMI_API, I have checked the article so I added this change and decided to notify about it.

**Comment by dcherniev** [comment]
Rebuild required

**Comment by dcherniev** [comment]
Rebuild required

**Comment by dcherniev** [comment]
Rebuild required

File: src/components/application_manager/include/application_manager/smart_object_keys.h [hunk 3]
```diff
@@ -216,6 +216,7 @@ extern const char* key_press_mode;
 // duplicate names from hmi_request
 extern const char* limited_character_list;
 extern const char* auto_complete_text;
+extern const char* auto_complete_list;
```
Comment (dtrunov) [inline src/components/application_manager/include/application_manager/smart_object_keys.h]: @Why did you add this parameter twice?

File: src/components/application_manager/src/commands/mobile/set_global_properties_request.cc [hunk 4]
```diff
@@ -393,7 +394,19 @@ void SetGlobalPropertiesRequest::PrepareUIRequestMenuAndKeyboardData(
...
+    const bool is_auto_complete_text_present =
+        out_params[hmi_request::keyboard_properties].keyExists(
+            strings::auto_complete_text);
+    const bool is_auto_complete_list_present =
+        out_params[hmi_request::keyboard_properties].keyExists(
+            strings::auto_complete_list);
+    if (is_auto_complete_list_present && is_auto_complete_text_present) {
+      out_params[hmi_request::keyboard_properties].erase(
+          strings::auto_complete_text);
+    }
```
Comment (dtrunov) [inline src/components/application_manager/src/commands/mobile/set_global_properties_request.cc]: @dcherniev Is it correct you checked two parameters from smart object and removed only one.

File: src/components/application_manager/src/smart_object_keys.cc [hunk 5]
```diff
@@ -182,6 +182,7 @@ const char* key_press_mode = "keypressMode";
 // duplicate names from hmi_request
 const char* limited_character_list = "limitedCharacterList";
 const char* auto_complete_text = "autoCompleteText";
+const char* auto_complete_list = "autoCompleteList";
 const char* navigation_text = "navigationText";
```
Comment (dtrunov) [inline src/components/application_manager/src/smart_object_keys.cc]: @dcherniev  Please leave one string

**Comment by dtrunov** [comment]
@dcherniev I think you need update resumption DB with new parameter, please check this

File: src/components/application_manager/src/commands/mobile/set_global_properties_request.cc (hunk 4 above)
Comment (dcherniev) [inline src/components/application_manager/src/commands/mobile/set_global_properties_request.cc]: @dtrunov According to requirements if parameters contain both (auto_complete_text and auto_complete_list) fields, the auto_complete_text field should be ommitted and only auto_complete_list field should be passed.

File: src/components/application_manager/include/application_manager/smart_object_keys.h (hunk 3 above)
Comment (dcherniev) [inline src/components/application_manager/include/application_manager/smart_object_keys.h]: @dtrunov I am a bit confused why there is a group of fields "duplicate names from hmi_request" which is duplicated in "namespace strings". Can you tell me in which namespase should I leave the new field "namespace hmi_request" or "namespace strings".

File: src/components/application_manager/src/smart_object_keys.cc (hunk 5 above)
Comment (dcherniev) [inline src/components/application_manager/src/smart_object_keys.cc]: @dtrunov This is the same situation that is in smart_object_keys.h file.

File: src/components/application_manager/include/application_manager/smart_object_keys.h (hunk 3 above)
Comment (dtrunov) [inline src/components/application_manager/include/application_manager/smart_object_keys.h]: @dcherniev You can use whether, first or second. I don't remember any agreements about these namespaces.

**Comment by dcherniev** [comment]
Rebuild required

File: src/components/application_manager/test/commands/mobile/set_global_properties_test.cc [hunk 6]
```diff
@@ -147,80 +193,69 @@ class SetGlobalPropertiesRequestTest
...
   EXPECT_CALL(app_mngr_, RemoveAppFromTTSGlobalPropertiesList(kConnectionKey));
   const smart_objects::SmartObject* vr_help_title =
-      &((*msg_vr)[am::strings::msg_params][am::strings::vr_help_title]);
+      &((*msg_tts)[strings::msg_params][strings::vr_help_title]);
   const smart_objects::SmartObject* vr_help =
-      &((*msg_vr)[am::strings::msg_params][am::strings::vr_help]);
+      &((*msg_tts)[strings::msg_params][strings::vr_help]);
   const smart_objects::SmartObject* vr_help_prompt =
-      &((*msg_vr)[am::strings::msg_params][am::strings::help_prompt]);
+      &((*msg_tts)[strings::msg_params][strings::help_prompt]);
```
Comment (LevchenkoS) [inline src/components/application_manager/test/commands/mobile/set_global_properties_test.cc]: @dcherniev, in my opinion, there is better to create a reference to `msg_params` like 
```C++
const smart_objects::SmartObject& msg_tts_params = (*msg_tts)[strings::msg_params];
// And then use it
const smart_objects::SmartObject* vr_help_title = 
    &(msg_tts_params[string::vr_help_title]);
const smart_objects::SmartObject* vr_help = 
    &(msg_tts_params[string::vr_help]);
const smart_objects::SmartObject* vr_help_prompt = 
    &(msg_tts_params[string::vr_help_prompt]);
```

File: src/components/application_manager/test/commands/mobile/set_global_properties_test.cc [hunk 7]
```diff
@@ -266,17 +295,54 @@ TEST_F(SetGlobalPropertiesRequestTest,
...
+
+  test_so[6][strings::msg_params][strings::menu_title] = kNotValidString;
+
+  test_so[7][strings::msg_params][strings::keyboard_properties]
+         [strings::limited_character_list][0] = kNotValidString;
+
+  test_so[8][strings::msg_params][strings::keyboard_properties]
+         [strings::auto_complete_text] = kNotValidString;
+
+  test_so[9][strings::msg_params][strings::keyboard_properties]
```
Comment (LevchenkoS) [inline src/components/application_manager/test/commands/mobile/set_global_properties_test.cc]: @dcherniev, may we abbreviate `string` namespace to `str`?

File: src/components/application_manager/src/commands/mobile/set_global_properties_request.cc (hunk 4 above)
Comment (dcherniev) [inline src/components/application_manager/src/commands/mobile/set_global_properties_request.cc]: @LuxoftAKutsan, @dtrunov Ok I would prefer to remove my new field from "duplicate names from hmi_request" however I think that this group of duplicate fields is a bit confusing, maybe I should delete the rest two fields in it too? (I can make it as a separate commit)

File: src/components/application_manager/test/commands/mobile/set_global_properties_test.cc (hunk 7 above)
Comment (dcherniev) [inline src/components/application_manager/test/commands/mobile/set_global_properties_test.cc]: @LevchenkoS Well I can, but don't we usually use the "namespace strings = application_manager::strings" abbreviation for strings?

File: src/components/application_manager/include/application_manager/smart_object_keys.h (hunk 3 above)
Comment (dcherniev) [inline src/components/application_manager/include/application_manager/smart_object_keys.h]: @LuxoftAKutsan, @dtrunov Ok I would prefer to remove my new field from "duplicate names from hmi_request" however I think that this group of duplicate fields is a bit confusing, maybe I should delete the rest two fields in it too? (I can make it as a separate commit)

Comment (dcherniev) [inline src/components/application_manager/include/application_manager/smart_object_keys.h]: @LuxoftAKutsan, @dtrunov Fixed in b10f1c6, also added TODO to check all other duplicate fields and leave only one exemplar for each of those fealds.

File: src/components/application_manager/test/commands/mobile/set_global_properties_test.cc (hunk 6 above)
Comment (dcherniev) [inline src/components/application_manager/test/commands/mobile/set_global_properties_test.cc]: @LevchenkoS Fixed,please check c0984ba.

File: src/components/application_manager/test/commands/mobile/set_global_properties_test.cc (hunk 7 above)
Comment (dcherniev) [inline src/components/application_manager/test/commands/mobile/set_global_properties_test.cc]: @LevchenkoS Fixed, used local namespace for this method, please check c0984ba.

File: src/components/application_manager/src/commands/mobile/set_global_properties_request.cc [hunk 8]
```diff
@@ -551,6 +564,26 @@ bool SetGlobalPropertiesRequest::IsWhiteSpaceExist() {
...
+            strings::auto_complete_list)) {
+      const smart_objects::SmartArray* acl_array =
+          msg_params[strings::keyboard_properties][strings::auto_complete_list]
+              .asArray();
+
+      smart_objects::SmartArray::const_iterator it_acl = acl_array->begin();
+      smart_objects::SmartArray::const_iterator it_acl_end = acl_array->end();
+
+      for (; it_acl != it_acl_end; ++it_acl) {
+        str = (*it_acl).asCharArray();
```
Comment (dcherniev) [inline src/components/application_manager/src/commands/mobile/set_global_properties_request.cc]: @LuxoftAKutsan Fixed, please check 871f38e.

**Comment by dcherniev** [comment]
@dtrunov I have clarified question about resumption DB, the changes related to this class are requested in another CRQ and are out of scope for this one.

File: src/components/application_manager/src/commands/mobile/set_global_properties_request.cc [hunk 9]
```diff
@@ -443,6 +455,7 @@ bool SetGlobalPropertiesRequest::IsWhiteSpaceExist() {
     smart_objects::SmartArray::const_iterator it_hp = hp_array->begin();
     smart_objects::SmartArray::const_iterator it_hp_end = hp_array->end();
 
+    const char* str;
```
Comment (LuxoftAKutsan) [inline src/components/application_manager/src/commands/mobile/set_global_properties_request.cc]: @dcherniev init it with NULL value.

File: src/components/application_manager/src/commands/mobile/set_global_properties_request.cc [hunk 10]
```diff
@@ -459,6 +472,7 @@ bool SetGlobalPropertiesRequest::IsWhiteSpaceExist() {
     smart_objects::SmartArray::const_iterator it_tp = tp_array->begin();
     smart_objects::SmartArray::const_iterator it_tp_end = tp_array->end();
 
+    const char* str;
```
Comment (LuxoftAKutsan) [inline src/components/application_manager/src/commands/mobile/set_global_properties_request.cc]: @dcherniev scope of htis var can be reduced even more.

File: src/components/application_manager/src/commands/mobile/set_global_properties_request.cc [hunk 11]
```diff
@@ -475,6 +489,7 @@ bool SetGlobalPropertiesRequest::IsWhiteSpaceExist() {
     smart_objects::SmartArray::const_iterator it_vh = vh_array->begin();
     smart_objects::SmartArray::const_iterator it_vh_end = vh_array->end();
 
+    const char* str;
```
Comment (LuxoftAKutsan) [inline src/components/application_manager/src/commands/mobile/set_global_properties_request.cc]: @dcherniev scope of htis var can be reduced even more.

File: src/components/application_manager/src/commands/mobile/set_global_properties_request.cc [hunk 12]
```diff
@@ -551,6 +568,27 @@ bool SetGlobalPropertiesRequest::IsWhiteSpaceExist() {
...
+    if (msg_params[strings::keyboard_properties].keyExists(
+            hmi_request::auto_complete_list)) {
+      const smart_objects::SmartArray* acl_array =
+          msg_params[strings::keyboard_properties]
+                    [hmi_request::auto_complete_list].asArray();
+
+      smart_objects::SmartArray::const_iterator it_acl = acl_array->begin();
+      smart_objects::SmartArray::const_iterator it_acl_end = acl_array->end();
+
+      const char* str;
```
Comment (LuxoftAKutsan) [inline src/components/application_manager/src/commands/mobile/set_global_properties_request.cc]: @dcherniev scope of htis var can be reduced even more.

File: src/components/application_manager/src/commands/mobile/set_global_properties_request.cc (hunk 9 above)
Comment (dcherniev) [inline src/components/application_manager/src/commands/mobile/set_global_properties_request.cc]: @LuxoftAKutsan Please check 45877df.

File: src/components/application_manager/src/commands/mobile/set_global_properties_request.cc (hunk 10 above)
Comment (dcherniev) [inline src/components/application_manager/src/commands/mobile/set_global_properties_request.cc]: @LuxoftAKutsan Fixed, please check 45877df.

File: src/components/application_manager/src/commands/mobile/set_global_properties_request.cc (hunk 11 above)
Comment (dcherniev) [inline src/components/application_manager/src/commands/mobile/set_global_properties_request.cc]: @LuxoftAKutsan Fixed, please check 45877df.

File: src/components/application_manager/src/commands/mobile/set_global_properties_request.cc (hunk 12 above)
Comment (dcherniev) [inline src/components/application_manager/src/commands/mobile/set_global_properties_request.cc]: @LuxoftAKutsan Fixed, please check 45877df.

File: src/components/application_manager/test/commands/mobile/set_global_properties_test.cc [hunk 13]
```diff
@@ -114,11 +128,11 @@ class SetGlobalPropertiesRequestTest
   MessageSharedPtr PrepareResponseFromHMI(
       const hmi_apis::Common_Result::eType result_code, const char* info) {
     MessageSharedPtr msg = CreateMessage(smart_objects::SmartType_Map);
-    (*msg)[am::strings::params][am::hmi_response::code] = result_code;
-    (*msg)[am::strings::msg_params] =
+    (*msg)[strings::params][am::hmi_response::code] = result_code;
```
Comment (dtrunov) [inline src/components/application_manager/test/commands/mobile/set_global_properties_test.cc]: @dcherniev Please remove redundant am from [am::hmi_response::code]

File: src/components/application_manager/test/commands/mobile/set_global_properties_test.cc [hunk 14]
```diff
@@ -114,11 +128,11 @@ class SetGlobalPropertiesRequestTest
...
+    (*msg)[strings::msg_params] =
```
Comment (dtrunov) [inline src/components/application_manager/test/commands/mobile/set_global_properties_test.cc]: @dcherniev Please remove am from [am::hmi_response::code]

**Comment by dcherniev** [comment]
@TMelnyk, please review my updated changes

**Comment by TMelnyk** [comment]
reviewed
